posthog/special_migrations/runner.py

```diff
@@ -0,0 +1,220 @@
+from typing import Optional, Tuple
+
+from semantic_version.base import SimpleSpec
+
+from posthog.models.special_migration import MigrationStatus, SpecialMigration, get_all_running_special_migrations
+from posthog.models.utils import UUIDT
+from posthog.special_migrations.setup import (
+    POSTHOG_VERSION,
+    get_special_migration_definition,
+    get_special_migration_dependency,
+)
+from posthog.special_migrations.utils import (
+    complete_migration,
+    execute_op,
+    process_error,
+    reset_special_migration,
+    trigger_migration,
+    update_special_migration,
+)
+
+"""
+Important to prevent us taking up too many celery workers and also to enable running migrations sequentially
+"""
+MAX_CONCURRENT_SPECIAL_MIGRATIONS = 1
+
+
+def start_special_migration(migration_name: str) -> bool:
+    """
+    Performs some basic checks to ensure the migration can indeed run, and then kickstarts the chain of operations
+    Checks:
+    1. We're not over the concurrent migrations limit
+    2. The migration can be run with the current PostHog version
+    3. The migration is not already running
+    4. The migration is required given the instance configuration
+    5. The service version requirements are met (e.g. X < ClickHouse version < Y)
+    6. The migration's healthcheck passes
+    7. The migration's dependency has been completed
+    """
+
+    migration_instance = SpecialMigration.objects.get(name=migration_name)
+    over_concurrent_migrations_limit = len(get_all_running_special_migrations()) >= MAX_CONCURRENT_SPECIAL_MIGRATIONS
+    if (
+        not migration_instance
+        or over_concurrent_migrations_limit
+        or not is_migration_in_range(migration_instance.posthog_min_version, migration_instance.posthog_max_version)
+        or migration_instance.status == MigrationStatus.Running
+    ):
+        return False
+
+    migration_definition = get_special_migration_definition(migration_name)
+
+    if not migration_definition.is_required():
+        complete_migration(migration_instance)
+        return False
+
+    for service_version_requirement in migration_definition.service_version_requirements:
+        in_range, version = service_version_requirement.is_service_in_accepted_version()
+        if not in_range:
+            process_error(
+                migration_instance,
+                f"Service {service_version_requirement.service} is in version {version}. Expected range: {str(service_version_requirement.supported_version)}.",
+            )
+
+    ok, error = run_migration_healthcheck(migration_instance)
+    if not ok:
+        process_error(migration_instance, error)
+        return False
+
+    dependency_ok, dependency_name = is_migration_dependency_fulfilled(migration_instance.name)
+
+    if not dependency_ok:
+        process_error(migration_instance, f"Could not trigger migration because it depends on {dependency_name}")
+        return False
+
+    reset_special_migration(migration_instance)
```

Comment:
First pass: semantics of "resetting" a migration are unclear w/o reading further.

Reply:
good point